### ISSUE TITLE
Expose some functions of frame-executive for building a custom executive easier

### ISSUE THREAD

**Issue by liuchengxu** [paritytech/substrate#10711]
A custom executive is required in Subspace, specifically, we would like to have a hook right after `initialize_block` and `apply_extrinsic` to collect the intermediate storage root during the entire block execution.

One idea is to create a wrapped pallet-executive on frame-executive, only copy a few functions(basically `execute_block` because it ties the initialize_block and extrinsic execution tightly) and make some modifications there, delegating all the other functions to frame-executive directly to minimize the maintenance of this custom pallet-executive. These functions just need to be made public in that way:

https://github.com/paritytech/substrate/blob/48e0b415b94d8ce14e0ebbbdb813263afa27f02e/frame/executive/src/lib.rs#L336
https://github.com/paritytech/substrate/blob/48e0b415b94d8ce14e0ebbbdb813263afa27f02e/frame/executive/src/lib.rs#L411
https://github.com/paritytech/substrate/blob/48e0b415b94d8ce14e0ebbbdb813263afa27f02e/frame/executive/src/lib.rs#L468

The another way could be some generic solution that is able to make the executive naturally more flexible to these custom needs, I haven't got a pretty clear idea though :P.

**Comment by liuchengxu** [comment]
@bkchr Any thoughts on this feature request?

**Comment by bkchr** [comment]
> to collect the intermediate storage root during the entire block execution.

Collect and then? What you want to do with that?

What happens at block production?

**Comment by nazar-pc** [comment]
Normally state root is collected after block execution, we also collect additional state roots after each transaction to make fraud proofs more efficient in our protocol.

**Comment by bkchr** [comment]
And you store them in the state?

**Comment by nazar-pc** [comment]
Currently yes

**Comment by bkchr** [comment]
I just have skimmed your code and the only change you made is the storage root calculation after `apply_extrinsic`. Is that true?

If yes, adding some callback there should be enough for you?

**Comment by liuchengxu** [comment]
>I just have skimmed your code and the only change you made is the storage root calculation after apply_extrinsic. Is that true?

Nope. Not only after `apply_extrinsic`, but also after `initialize_block` https://github.com/subspace/subspace/blob/438c569b38/cumulus/pallets/executive/src/lib.rs#L238.

>adding some callback there should be enough for you?

Yeah, that should be enough, I need two hooks to run right after `initiali_block` and `apply_extrinsic`.

**Comment by bkchr** [comment]
Ahh okay, yeah than adding some callback to these both and maybe after calling `on_finalize` should be okay?

**Comment by bkchr** [comment]
WDYT @shawntabrizi

**Comment by liuchengxu** [comment]
Ping @shawntabrizi
